# Dimensions, Shapes, and Array Attributes

In NumPy to represent a 2D grid of numbers as an array of arrays we use multi-dimensional arrays.

* A 1D array is like a vector (a single list of numbers).
* A 2D array is a matrix (rows and columns, like a spreadsheet).
* A 3D array is a tensor (a collection of matrices, like layers of a cake or color channels of
    an image: Red, Green, and Blue).

When you feed data into machine learning models (like TensorFlow or PyTorch), they expect your data to have very specific dimensions and shapes. If your dimensions are wrong, the code will fail. Knowing how to check these attributes is the key to debugging and understanding your data.


#### 1D Array (Vector): A Street
A 1D array is like a single street with houses.
* To find a house, you only need **one address number** (e.g., House 3).
* This represents a single sequence of numbers.

#### 2D Array (Matrix): A Parking Lot
A 2D array is like a giant parking lot with **rows and columns**.
* To find a parked car, you need **two numbers**: the Row number and the Column number (e.g., Row 2, Column 4).

#### 3D Array (Tensor): A Multi-Story Parking Garage
A 3D array is like a parking garage with **multiple levels**.
* To find your car, you need **three numbers**: Level (depth), Row, and Column (e.g., Level 1, Row 2, Column 4).

#### Attributes Explained:
* `.ndim`: **How many numbers** do you need to find a single element? (1 for a street, 2 for a parking lot, 3 for a garage).
* `.shape`: **How big** is the street, parking lot, or garage? For a lot, shape might be `(3, 4)`: 3 rows and 4 columns. For a garage, it might be `(2, 3, 4)`: 2 levels, 3 rows, 4 columns.
* `.size`: **How many total spots** are there? In a 3-row, 4-column lot, there are `3 * 4 = 12` total spots.
* `.dtype`: **What type of vehicle** is allowed? (only trucks, or only sports cars). In NumPy, every spot in the array must hold the exact same type of data.


## Code Examples and Outputs

Let's create 1D, 2D, and 3D arrays and inspect their properties.

#### Example: Defining and Inspecting a 1D, 2D, and 3D Array

In [2]:
import numpy as np

# 1. 1D Array (a single row)
arr_1d = np.array([1, 2, 3, 4, 5])

# 2. 2D Array (rows and columns)
arr_2d = np.array([
    [1, 2, 3],
    [4, 5, 6]
])

# 3. 3D Array (2 pages/layers, each with a 2x3 matrix)
arr_3d = np.array([
    [[1, 2, 3], [4, 5, 6]],
    [[7, 8, 9], [10, 11, 12]]
])

print("--- arr_1d ---")
print("ndim :", arr_1d.ndim)
print("shape:", arr_1d.shape)
print("size :", arr_1d.size)

print("- -- arr_2d - --")
print("ndim :", arr_2d.ndim)
print("shape:", arr_2d.shape)
print("size :", arr_2d.size)

print("- -- arr_3d - --")
print("ndim :", arr_3d.ndim)
print("shape:", arr_3d.shape)
print("size :", arr_3d.size)

--- arr_1d ---
ndim : 1
shape: (5,)
size : 5
- -- arr_2d - --
ndim : 2
shape: (2, 3)
size : 6
- -- arr_3d - --
ndim : 3
shape: (2, 2, 3)
size : 12


**Why this works:**

* `arr_1d.shape` returns `(5,)` which is a tuple containing 1 number. This is how Python represents a 1D vector.
* `arr_2d.shape` is `(2, 3)` which means 2 rows and 3 columns.
* `arr_3d.shape` is `(2, 2, 3)` which means 2 layers, 2 rows, and 3 columns.
* `.size` is always the product of the numbers in the shape (e.g., `2 * 2 * 3 = 12`).

#### Example: Inspecting Data Types and Memory Usage

In [3]:
import numpy as np

# Create an array of 32-bit integers
int_arr = np.array([1, 2, 3], dtype=np.int32)

# Create an array of 64-bit floats
float_arr = np.array([1.0, 2.0, 3.0], dtype=np.float64)

print("int_arr dtype   :", int_arr.dtype)
print("int_arr itemsize:", int_arr.itemsize, "bytes per element")
print("int_arr nbytes  :", int_arr.nbytes, "total bytes in memory")

print("float_arr dtype   :", float_arr.dtype)
print("float_arr itemsize:", float_arr.itemsize, "bytes per element")
print("float_arr nbytes  :", float_arr.nbytes, "total bytes in memory")

int_arr dtype   : int32
int_arr itemsize: 4 bytes per element
int_arr nbytes  : 12 total bytes in memory
float_arr dtype   : float64
float_arr itemsize: 8 bytes per element
float_arr nbytes  : 24 total bytes in memory


**Why this works:**

* `.itemsize` is the length of one array element in bytes. `int32` uses 32 bits (4 bytes), so its itemsize is `4`.
* `.nbytes` is the total memory used by the elements in the array (`size * itemsize`). Here, `3 * 4 = 12` bytes for `int_arr` and `3 * 8 = 24` bytes for `float_arr`.

#### Example: Upcasting (Type Promotion)

What happens if we mix different types of data in a single array?

In [4]:
import numpy as np

# Mix integers and floats
mix_nums = np.array([1, 2.5, 3])
print("Mixed numbers dtype:", mix_nums.dtype)
print("Mixed numbers array:", mix_nums)

# Mix numbers and strings
mix_strings = np.array([1, "hello", 3.5])
print("Mixed strings dtype:", mix_strings.dtype)
print("Mixed strings array:", mix_strings)

Mixed numbers dtype: float64
Mixed numbers array: [1.  2.5 3. ]
Mixed strings dtype: <U32
Mixed strings array: ['1' 'hello' '3.5']


**Why this works:**

* An array must have a single data type.
* When mixing integers and floats, NumPy promotes (upcasts) the integers to floats because `2.5` cannot be represented as an integer, but `1` can easily be represented as a float `1.0`. This preserves your data precision.
* When mixing numbers and strings, NumPy converts **everything to strings** (represented as `<U32`, which means Unicode string of maximum 32 characters). This is because you cannot represent the word `"hello"` as a number! This makes your array unoptimized and slow. **Avoid mixing strings and numbers in NumPy arrays!**



#### Mistake 1: Creating an Inhomogeneous (Unbalanced) Shape

If you try to create a 2D array where the rows have different lengths, NumPy will throw an error or create a slow object array.


In [5]:
# WRONG: Row 1 has 3 elements, Row 2 has 2 elements!
try:
    bad_arr = np.array([[1, 2, 3], [4, 5]])
except ValueError as e:
    print("Error caught:", e)

Error caught: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (2,) + inhomogeneous part.



**Correction:** Every row must have the exact same number of elements.

In [ ]:
good_arr = np.array([[1, 2, 3], [4, 5, 0]]) # Use a placeholder like 0

#### Exercise 1
Create a 1D array from the list `[10, 20, 30]`.
1. Print its shape.
2. Print its ndim (number of dimensions).
3. Print its size (total elements).

In [6]:
import numpy as np

arr = np.array([10, 20, 30])
print("Shape:", arr.shape)
print("Ndim :", arr.ndim)
print("Size :", arr.size)

Shape: (3,)
Ndim : 1
Size : 3


**Explanation:** A 1D vector of 3 items has a shape tuple `(3,)`. It is 1-dimensional (`ndim = 1`) and has a size of 3.



#### Exercise 2

Create a 2D array with 3 rows and 4 columns containing any numbers of your choice.
1. Inspect and print its shape, ndim, and size.
2. Verify that its size is equal to the product of its shape dimensions (Rows * Columns).

In [7]:
import numpy as np

matrix = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12]
])

print("Shape:", matrix.shape)
print("Ndim :", matrix.ndim)
print("Size :", matrix.size)

# Verification
rows, cols = matrix.shape
print(f"Is size equal to rows * cols? {matrix.size == rows * cols}")

Shape: (3, 4)
Ndim : 2
Size : 12
Is size equal to rows * cols? True


**Explanation:** The matrix has 3 rows and 4 columns, so its shape is `(3, 4)`. The total number of elements is `3 * 4 = 12`.


#### Exercise 3

Create a 3D array of ones with a shape of `(2, 3, 4)` using a built-in NumPy function.

1. Print the array.
2. Print its `.ndim`, `.shape`, `.size`, `.dtype`, and `.nbytes`.
3. Change its `dtype` to `int16`. Print its `.dtype` and `.nbytes` again. Explain why `.nbytes` changed.


In [8]:
import numpy as np

# Create 3D array of ones
arr_3d = np.ones((2, 3, 4), dtype=np.float64) # Default float64
print("--- Float64 Array ---")
print("ndim :", arr_3d.ndim)
print("shape:", arr_3d.shape)
print("size :", arr_3d.size)
print("dtype:", arr_3d.dtype)
print("nbytes:", arr_3d.nbytes)

# Change dtype to int16 (using astype)
arr_3d_int = arr_3d.astype(np.int16)
print("--- Int16 Array ---")
print("dtype :", arr_3d_int.dtype)
print("nbytes:", arr_3d_int.nbytes)

--- Float64 Array ---
ndim : 3
shape: (2, 3, 4)
size : 24
dtype: float64
nbytes: 192
--- Int16 Array ---
dtype : int16
nbytes: 48


**Explanation:**

* The size is `2 * 3 * 4 = 24` elements.
* A `float64` element uses 64 bits (8 bytes) of memory. Total memory used is `24 * 8 = 192` bytes.
* An `int16` element uses 16 bits (2 bytes) of memory. Total memory used is `24 * 2 = 48` bytes.
* Changing the data type to a smaller bit size dramatically reduces the memory footprints of your array (from 192 bytes down to 48 bytes). This is extremely useful when working with very large datasets.
